### PR TITLE
[REVIEW] Raise temporary error for `decimal128` types in parquet reader

### PR DESCRIPTION
**Pull request by galipremsagar**
This PR adds a `decimal128` type validation in parquet reader. This is put in-place to unblock libcudf changes: https://github.com/rapidsai/cudf/pull/9765 and this validation will soon be removed once python side of `decimal128` changes are merged(blocked by libcudf `from_arrow` bug).

### CODE REVIEW

File: python/cudf/cudf/io/parquet.py [hunk 1]
```diff
@@ -411,6 +412,26 @@ def read_parquet(
             filepaths_or_buffers.append(tmp_source)
 
     if engine == "cudf":
+        # Temporary error to probe a parquet file
+        # and raise decimal128 support error.
+        if len(filepaths_or_buffers) > 0:
+            try:
+                metadata = pq.read_metadata(filepaths_or_buffers[0])
```
Comment (vuule) [inline python/cudf/cudf/io/parquet.py:419]: Is the schema flat?

Comment (galipremsagar) [reply python/cudf/cudf/io/parquet.py:419]: Pushed the changes to handle nested schema types too.

File: python/cudf/cudf/tests/test_parquet.py [hunk 2]
```diff
@@ -626,6 +626,7 @@ def test_parquet_reader_spark_timestamps(datadir):
     assert_eq(expect, got)
 
 
+@pytest.mark.xfail(reason="decimal128 not yet supported in cuDF")
```
Comment (vuule) [inline python/cudf/cudf/tests/test_parquet.py:629]: should we modify this test to expect a failure instead of adding a new file?

Comment (galipremsagar) [reply python/cudf/cudf/tests/test_parquet.py]: Updated the tests accordingly 👍

File: python/cudf/cudf/tests/test_parquet.py [hunk 3]
```diff
@@ -629,15 +629,28 @@ def test_parquet_reader_spark_timestamps(datadir):
...
-    expect = expect.apply(pd.to_numeric)
+    # expect = expect.apply(pd.to_numeric)
 
     # np.testing.assert_allclose(expect, got)
-    assert_eq(expect, got)
+    # assert_eq(expect, got)
+
+
+def test_parquet_reader_decimal128_error_validation(datadir):
+    fname = datadir / "nested_decimal128_file.parquet"
```
Comment (vuule) [inline python/cudf/cudf/tests/test_parquet.py:648]: what is the schema of this file?

Comment (galipremsagar) [reply python/cudf/cudf/tests/test_parquet.py:648]: This is the current schema:
```
>>> m.schema
<pyarrow._parquet.ParquetSchema object at 0x7f7a5dc52ec0>
required group field_id=-1 schema {
  optional group field_id=-1 a (List) {
    repeated group field_id=-1 list {
      optional fixed_len_byte_array(9) field_id=-1 item (Decimal(precision=20, scale=1));
    }
  }
}

```

Do you want the test to include some specific schema?

File: python/cudf/cudf/io/parquet.py [hunk 4]
```diff
@@ -731,3 +756,11 @@ def merge_parquet_filemetadata(filemetadata_list):
 
 
 ParquetWriter = libparquet.ParquetWriter
+
+
+def _check_decimal128_type(arrow_type):
+    if isinstance(arrow_type, pa.Decimal128Type):
+        if arrow_type.precision > cudf.Decimal64Dtype.MAX_PRECISION:
+            raise NotImplementedError(
+                "Decimal type greater than Decimal64 is not " "yet supported"
```
Comment (vuule) [inline python/cudf/cudf/io/parquet.py:765]:
```suggestion
                "Decimal type greater than Decimal64 is not yet supported"
```

**Review by vuule — APPROVED** [review]
Looks good, thanks for taking care of this!

File: python/cudf/cudf/io/parquet.py [hunk 5]
```diff
@@ -614,6 +615,30 @@ def _read_parquet(
     # Simple helper function to dispatch between
     # cudf and pyarrow to read parquet data
     if engine == "cudf":
+        # Temporary error to probe a parquet file
+        # and raise decimal128 support error.
+        if len(filepaths_or_buffers) > 0:
+            try:
+                metadata = pq.read_metadata(filepaths_or_buffers[0])
+            except TypeError:
```
Comment (vyasr) [inline python/cudf/cudf/io/parquet.py:623]: What is this branch handling? I always find it helpful to see comments in except clauses ignoring an error, it's very helpful in debugging subtle bugs down the line if the caught exception is too broad.

**Review by vyasr — COMMENTED** [review]
Do we know anything about the performance implications of this? I don't know how expensive functions like `to_arrow_schema` and `pq.read_metadata` are, but it does seem like significant error-checking work to be done up front. Would it be possible to try reading the file, catching whatever error we get back from that, and then checking for 128-bit columns after the fact?

File: python/cudf/cudf/io/parquet.py (hunk 5 above)
Comment (galipremsagar) [reply python/cudf/cudf/io/parquet.py:623]: Added a comment to explain the same.

**Comment by galipremsagar** [comment]
> Do we know anything about the performance implications of this? I don't know how expensive functions like `to_arrow_schema` and `pq.read_metadata` are, but it does seem like significant error-checking work to be done up front. Would it be possible to try reading the file, catching whatever error we get back from that, and then checking for 128-bit columns after the fact?

Here is a benchmark of reading 1000 columns dataframe:

```ipython
In [3]: df = cudf.DataFrame(columns=range(1, 1000))

In [4]: df
Out[4]: d
Empty DataFrame
Columns: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 980, 981, 982, 983, 984, 985, 986, 987, 988, 989, 990, 991, 992, 993, 994, 995, 996, 997, 998, 999]
Index: []

In [5]: df[1] = ['a', 'b', 'c']

In [6]: df
Out[6]: 
  1     2     3     4     5     6     7     8     9     10    11    12    13    14    15   ...   985   986   987   988   989   990   991   992   993   994   995   996   997   998   999
0   a  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  ...  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>
1   b  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  ...  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>
2   c  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  ...  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>  <NA>

[3 rows x 999 columns]

In [10]: df.columns = df.columns.astype('str')

In [11]: df.to_parquet('long.parquet')

# Before
In [12]: %timeit cudf.read_parquet("long.parquet")
151 ms ± 1.61 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

# This PR
In [3]: %timeit cudf.read_parquet("long.parquet")
171 ms ± 4.53 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

File: python/cudf/cudf/tests/test_parquet.py [hunk 6]
```diff
@@ -629,15 +629,29 @@ def test_parquet_reader_spark_timestamps(datadir):
...
     fname = datadir / "spark_decimal.parquet"
 
-    expect = pd.read_parquet(fname)
-    got = cudf.read_parquet(fname)
+    # expect = pd.read_parquet(fname)
+    with pytest.raises(
+        NotImplementedError,
+        match="Decimal type greater than Decimal64 is not yet supported",
+    ):
+        _ = cudf.read_parquet(fname)
```
Comment (vyasr) [inline python/cudf/cudf/tests/test_parquet.py:637]:
```suggestion
        cudf.read_parquet(fname)
```

File: python/cudf/cudf/tests/test_parquet.py [hunk 7]
```diff
@@ -629,15 +629,29 @@ def test_parquet_reader_spark_timestamps(datadir):
...
 
     # Convert the decimal dtype from PyArrow to float64 for comparison to cuDF
     # This is because cuDF returns as float64 as it lacks an equivalent dtype
-    expect = expect.apply(pd.to_numeric)
+    # expect = expect.apply(pd.to_numeric)
 
     # np.testing.assert_allclose(expect, got)
-    assert_eq(expect, got)
+    # assert_eq(expect, got)
```
Comment (vyasr) [inline python/cudf/cudf/tests/test_parquet.py:644]: Is the plan to uncomment these lines of code once Decimal128 support is properly added?

**Review by vyasr — COMMENTED** [review]
> I thought it was OK to introduce this overhead since this is a temporary one and will soon be gone once python support is enabled for Decimal128Column. But feel free to let me know your thoughts on this as I could be wrong too.

Thanks for the context, I missed that. A 10-15% performance regression is pretty undesirable, but not as big a deal if it's temporary. I think that as long as we anticipate the Decimal128 issues being resolved in this release it should be fine. I wouldn't want to release 22.02 with this performance regression in it though. Do you or @vuule know what the timeline is?

**Comment by galipremsagar** [comment]
>  I wouldn't want to release 22.02 with this performance regression in it though. Do you or @vuule know what the timeline is?

Yup, it is being targeted for 22.02

File: python/cudf/cudf/tests/test_parquet.py (hunk 7 above)
Comment (galipremsagar) [reply python/cudf/cudf/tests/test_parquet.py:644]: Yup

**Comment by vuule** [comment]
> A 10-15% performance regression is pretty undesirable, but not as big a deal if it's temporary

IMO it's fine, as this is 15% with an empty dataframe, and the overhead does not increase with the number of rows. So it should be negligible in most cases.

**Review by vyasr — APPROVED** [review]
>   A 10-15% performance regression is pretty undesirable, but not as big a deal if it's temporary
> IMO it's fine, as this is 15% with an empty dataframe, and the overhead does not increase with the number of rows. So it should be negligible in most cases.
It's not actually empty, right? @galipremsagar added enough data that it will actually write out a meaningful file with many columns, so it's not a trivial read. If the overhead doesn't scale with the number of rows then I agree that it should become negligible for a reasonable size though.

In any case, as long as we're targeting this for 22.02 the discussion is moot and I'm fine with this solution.

**Comment by galipremsagar** [comment]
@gpucibot merge

**Pull request merged**